Adds documentation for the FixtureLifeCycle attribute (#535)
nunit/docs: docs/articles/nunit/writing-tests/attributes/fixturelifecycle.md

---
uid: fixturelifecycleattribute
---

# FixtureLifeCycle

> Added in **NUnit 3.13.**

Author: I'd like to start adding this for any new feature that is added after the 3.0 release so users know which version the documentation applies to. If anyone has a better idea of how to format this, please post suggestions.

Reviewer: It looks very nice!

Reviewer 2: This LGTM 👍 . Ideally we'd have different docs for various releases, but at the moment traceability like this is our best option, and will help us in the future.

Author: I wouldn't mind a dropdown at the top of each page that says what release it is for, but a separate set of docs became a bit of a problem with the NUnit 2.x docs. People linked to the current release at the time, so Google searches ended up pointing to old documentation and people ended up missing updates in newer docs.

> The **FixtureLifeCycleAttribute** is used to indicate that an instance for a test fixture or all test fixtures in an assembly should be constructed for each test within the fixture or assembly.

Reviewer 2: ```suggestion
The `FixtureLifeCycleAttribute` is used to indicate that an instance for a test fixture or all test fixtures in an assembly should be constructed for each test within the fixture or assembly.
```

Suggest using backticks rather than bold as this has some implications for our spell-checking.

This attribute may be applied to a test fixture (class) or to a test assembly. It is useful in combination with the [Parallelizable Attribute](parallelizable.md) so that a new instance of a test fixture is constructed for every test within the test fixture. This allows tests to run in isolation without sharing instance fields and properties during parallel test runs. This make running parallel tests easier because it is easier to make your tests thread safe.

> This attribute can be applied to Classes or to the entire test Assembly. If applied to an Assembly, it may be overridden at the class level.

Reviewer: I would not capitalize 'class' or 'assembly' because they are not functioning as proper nouns in the sentence.

> This attribute can be applied to Classes or to the entire test Assembly. If applied to an Assembly, it may be overridden at the class level.

Reviewer 2: ```suggestion
This attribute can be applied to classes or to the entire test assembly. If applied to an assembly, it may be overridden at the class level.
```

Codifying @jnm2's suggestions so that you can accept them directly if you agree.

## LifeCycle Enumeration

> The constructor of **FixtureLifeCycleAttribute** takes a **LifeCycle** attribute to indicate if a single instance of a test fixture should be created for all tests or if a new instance should be created for each test.

Reviewer 2: ```suggestion
The constructor of `FixtureLifeCycleAttribute` takes a `LifeCycle` attribute to indicate if a single instance of a test fixture should be created for all tests or if a new instance should be created for each test.
```

Suggest backticks vs bolding.

>  Value | Meaning
> -------|---------
> **LifeCycle.SingleInstance**     | A single instance is created and shared for all test cases
> **LifeCycle.InstancePerTestCase** | A new instance is created for each test case

Reviewer 2: ```suggestion
| Value | Meaning |
| -------|--------- |
| `LifeCycle.SingleInstance` | A single instance is created and shared for all test cases |
| `LifeCycle.InstancePerTestCase` | A new instance is created for each test case |
```

Suggesting table formatting consistent with elsewhere in the docs. Also suggesting backticks vs bolding.

## Notes

> 1. When using **LifeCycle.InstancePerTestCase**, **OneTimeSetUp** and **OneTimeTearDown** methods must be static and each are only called once. This is required so that the setup or teardown methods do not access instance fields or properties that are reset for every test.

Reviewer 2: ```suggestion
* When using `LifeCycle.InstancePerTestCase`, the [`OneTimeSetUp`](xref:onetimesetup-attribute) and [`OneTimeTearDown`](xref:onetimeteardown-attribute) methods must be static, and each are only called once. This is required so that the setup or teardown methods do not access instance fields or properties that are reset for every test.
```

Suggesting bullet points, backticks, and xrefs to the other attributes.

> 2. When using **LifeCycle.InstancePerTestCase**, a class's constructor will be called before every test is executed and **IDisposable** test fixtures will be disposed after the test is finished.

Reviewer 2: ```suggestion
* When using `LifeCycle.InstancePerTestCase`, a class's constructor will be called before every test is executed and `IDisposable` test fixtures will be disposed after the test is finished.
```

Suggesting bullets and backticks

> 3. **SetUp** and **TearDown** methods are called before and after every test.

Reviewer 2: ```suggestion
* [`SetUp`](xref:setup-attribute) and [`TearDown`](xref:teardown-attribute) methods are called before and after every test.
```

Suggesting bullets, backticks, and xrefs.

> 4. The **Order** attribute is respected.

Reviewer 2: ```suggestion
* The `Order` attribute is respected.
```

Suggesting bullet and backtick here. Not suggesting xref as for some reason the order attribute doesn't have a uid. I'll fix that in a quick PR.

## See Also

> * [Parallelizable Attribute](parallelizable.md)
> * [OneTimeSetUp Attribute](onetimesetup.md)
> * [OneTimeTearDown Attribute](onetimeteardown.md)

Reviewer 2: ```suggestion
* [Parallelizable Attribute](xref:parallelizableattribute)
* [OneTimeSetUp Attribute](xref:onetimesetup-attribute)
* [OneTimeTearDown Attribute](xref:onetimeteardown-attribute)
```
Suggesting the use of xrefs in case the document structure changes at some point. Docfx allows you to use the `xref` syntax to reference a document's `uid`.
